src/charm.py

```diff
@@ -0,0 +1,210 @@
+            basic_auth_password=self.config["prometheus_basic_auth_password"],
+            extra_labels=self.config["prometheus_extra_labels"],
+            metrics_path=self.config["prometheus_metrics_path"],
+        )
+        cfg = ServiceConfig(
+            ip=self.private_ip,
+            port=int(self.config["port"]),
+            log_level=self.config["log_level"],
+            metrics=metrics_cfg,
+            backend=backend_cfg,
+            store=etcd_cfg,
+        )
+        self._snap.configure(cfg)
+        if self.config["location"]:
+            self._snap.set_location(self.config["location"], self.config["port"])
+        self.unit.set_ports(int(self.config["port"]))
+        self.unit.status = ActiveStatus()
+
+    def _on_lxd_integrator_joined(self, event: RelationJoinedEvent):
+        cert, key = self._generate_selfsigned_cert(self.public_ip, self.public_ip, self.private_ip)
+        self._snap.setup_lxd(cert=cert, key=key)
+        relation_data = event.relation.data[self.unit]
+        relation_data["client_certificates"] = json.dumps([cert.decode("utf-8")])
+
+    def _generate_selfsigned_cert(self, hostname, public_ip, private_ip) -> tuple[bytes, bytes]:
+        if not hostname:
+            raise Exception("A hostname is required")
+
+        if not public_ip:
+            raise Exception("A public IP is required")
+
+        if not private_ip:
+            raise Exception("A private IP is required")
+
+        ca_key = generate_private_key(key_size=4096)
+        ca_cert = generate_ca(ca_key, hostname)
+
+        key = generate_private_key(key_size=4096)
+        csr = generate_csr(
+            private_key=key,
+            subject=hostname,
+            sans_dns=[public_ip, private_ip, hostname],
+            sans_ip=[public_ip, private_ip],
+        )
+        cert = generate_certificate(csr=csr, ca=ca_cert, ca_key=ca_key)
+        return cert, key
+
+    def _on_rest_api_joined(self, event: RelationJoinedEvent):
+        remote_data = event.relation.data.get(event.unit)
+        if not remote_data:
+            event.defer()
+            return
+        client_cert = remote_data.get("client_certificate")
+        if not client_cert:
+            event.defer()
+            logger.error("No client certificate found")
+            return
+        fingerprint = self._snap.register_client(ast.literal_eval(client_cert))
+        if fingerprint:
+            self._state.registered_clients.add(f"{event.unit.name}:{fingerprint}")
+        logger.info(self._state.registered_clients)
+        if not self._snap.is_running:
+            event.defer()
+            return
+
+        data = {
+            "port": str(self.config["port"]),
+            "private_address": self.private_ip,
+            "public_address": self.public_ip,
+            "node": self.unit.name.replace("/", ""),
+        }
+        location = self._snap.get_config_item("load_balancer.url")
```

Comment:
This reads odd and lets me ask my self "why do we retrieve a config item from the snap?", then I see that  `self._snap` is of `class AMS` and it makes sense. Can we rename `self._snap` to `self._ams` or similar to reflect that? Otherwise this expresses that we perform an operation on the snap (which we don't do, we call `amc config show`).

Reply:
This is covered in my next PR 😄